File: l2_engine/pnl_analyzer.py
# l2_engine/pnl_analyzer.py
import json

class PnLAnalyzer:
    @staticmethod
    def calculate_attribution(holdings, current_map):
        """
        holdings: [{"ticker": "0700.HK", "cost": 300, "quantity": 100}, ...]
        current_map: {"0700.HK": {"price": 320, "change_percent": ...}, ...}
        Note: current_map structure matches MarketMCP output or just simple price dict
        """
        report = []
        total_pnl = 0
        total_cost = 0
        
        for item in holdings:
            ticker = item['ticker']
            
            # 兼容 current_map 是 float 或者是 dict 的情况
            market_data = current_map.get(ticker, {})
            if isinstance(market_data, (int, float)):
                curr_price = float(market_data)
            else:
                curr_price = float(market_data.get('current_price', item['cost']))
            
            cost = item['cost']
            quantity = item['quantity']
            
            pnl = (curr_price - cost) * quantity
            total_pnl += pnl
            total_cost += cost * quantity
            
            return_rate = ((curr_price / cost) - 1) * 100 if cost != 0 else 0
            
            report.append({
                "ticker": ticker,
                "current_price": round(curr_price, 2),
                "cost_price": round(cost, 2),
                "quantity": quantity,
                "pnl": round(pnl, 2),
                "return_rate": f"{return_rate:.2f}%"
            })
            
        total_return_rate = (total_pnl / total_cost * 100) if total_cost != 0 else 0

        return {
            "total_pnl": round(total_pnl, 2),
            "total_return_rate": f"{total_return_rate:.2f}%",
            "details": report,
            "summary": "盈利" if total_pnl > 0 else "亏损"
        }

File: src/l2_engine/pnl_analyzer.ts
import { Holding, MarketData, PnLReport, PnLDetail } from '../types';

/**
 * L2 Engine - PnL Analyzer
 * 盈亏分析引擎，计算持仓盈亏和归因
 */
export class PnLAnalyzer {
  /**
   * 计算持仓盈亏归因
   * @param holdings 持仓列表
   * @param currentMap 当前市场价格映射
   * @returns 盈亏分析报告
   */
  static calculateAttribution(
    holdings: Holding[],
    currentMap: Map<string, MarketData | number>
  ): PnLReport {
    const report: PnLDetail[] = [];
    let totalPnl = 0;
    let totalCost = 0;

    for (const item of holdings) {
      const ticker = item.ticker;
      const marketData = currentMap.get(ticker);

      // 兼容 current_map 是 number 或者是 object 的情况
      let currPrice: number;
      if (typeof marketData === 'number') {
        currPrice = marketData;
      } else if (marketData && typeof marketData === 'object') {
        currPrice = marketData.current_price ?? item.cost;
      } else {
        currPrice = item.cost;
      }

      const cost = item.cost;
      const quantity = item.quantity;

      const pnl = (currPrice - cost) * quantity;
      totalPnl += pnl;
      totalCost += cost * quantity;

      const returnRate = cost !== 0 ? ((currPrice / cost) - 1) * 100 : 0;

      report.push({
        ticker,
        current_price: Math.round(currPrice * 100) / 100,
        cost_price: Math.round(cost * 100) / 100,
        quantity,
        pnl: Math.round(pnl * 100) / 100,
        return_rate: `${returnRate.toFixed(2)}%`,
      });
    }

    const totalReturnRate = totalCost !== 0 ? (totalPnl / totalCost) * 100 : 0;

    return {
      total_pnl: Math.round(totalPnl * 100) / 100,
      total_return_rate: `${totalReturnRate.toFixed(2)}%`,
      details: report,
      summary: totalPnl > 0 ? '盈利' : '亏损',
    };
  }
}
